In [123]:
import numpy as np
import pandas as pd
import datetime as dt

# Q1

### a) The file "Purchase_Data.csv" contains purchase data at a Chinese supermarket for various products.  Read this csv into a dataframe called df_data and use the head() method to check that you have read in the data correctly. Use dtypes to view column datatypes.  <br>  (Mark: 1)

In [124]:
df_data = pd.read_csv("Purchase_Data.csv")

df_data.head()

invoice_number   invoice_time store_id  product_id  category_id  num_item  \
0    2.020000e+11  7/15/15 10:45     G816     7000300       210101       1.0   
1    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
2    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
3    2.020000e+11  7/15/15 12:16     G816     7070891        70401       1.0   
4    2.020000e+11  7/15/15 12:16     G816     7090363        70401       1.0   

   price  
0    2.0  
1    2.0  
2    2.0  
3    6.0  
4    6.0

In [125]:
df_data.dtypes

invoice_number    float64
invoice_time       object
store_id           object
product_id          int64
category_id         int64
num_item          float64
price             float64
dtype: object

### b) Slice the first two columns and first two rows using .loc and store the result in a variable called result_1. <br>(Mark: 1)

In [126]:
result_1 = df_data.loc[0:1, 'invoice_number':'invoice_time']
result_1

invoice_number   invoice_time
0    2.020000e+11  7/15/15 10:45
1    2.020000e+11  7/15/15 12:16

### c) Select rows 0,4,6 and columns 'invoice time' and 'price' and store the result in variable called result_2. <br>(Mark: 1)

In [127]:
result_2 = df_data.iloc[[0,4,6], [1,6]]
result_2

invoice_time  price
0  7/15/15 10:45    2.0
4  7/15/15 12:16    6.0
6  7/15/15 15:28    2.0

### d) Store the number rows in a variable called num_rows. Extract the last row of the data, using  the variable num_rows and loc. Show the price of that row. <br>(Mark: 1)

In [128]:
df_data.shape[0]

5382

In [129]:
num_rows = df_data.shape[0]

df_data.loc[num_rows-1:, 'price']

5381    2.0
Name: price, dtype: float64

### e) Compute the average and max price charged.  Store the results in variables called avg_price and max_price <br> (Mark: 1) 

In [130]:
avg_price = round(df_data['price'].mean(),2)
print("average price charged = $", avg_price)

max_price = df_data['price'].max()
print("max price charged = $", max_price)

average price charged = $ 3.19
max price charged = $ 6.01


### f) Create a column called "total", which is the product of the price and num_item columns. <br> (Mark: 1)

In [131]:
df_data['total'] = df_data.num_item * df_data.price

df_data.head(20)

invoice_number   invoice_time store_id  product_id  category_id  num_item  \
0     2.020000e+11  7/15/15 10:45     G816     7000300       210101       1.0   
1     2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
2     2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
3     2.020000e+11  7/15/15 12:16     G816     7070891        70401       1.0   
4     2.020000e+11  7/15/15 12:16     G816     7090363        70401       1.0   
5     2.020000e+11  7/15/15 14:10     G816     7000300       210101       1.0   
6     2.020000e+11  7/15/15 15:28     G816     7000126       210301       1.0   
7     2.020000e+11  7/15/15 16:20     G816     7308017       210101       1.0   
8     2.020000e+11  7/15/15 16:22     G816    15000231        70301       1.0   
9     2.020000e+11  7/15/15 16:24      628     7000300       210101       1.0   
10    2.020000e+11  7/15/15 16:25     G816     7070884        70203       3.0   
11    2.020000e+11  7/15/15 16:39     G816     7070884        70203       1.0   
12    2.020000e+11  7/15/15 16:39     G816     7070884        70203       1.0   
13    2.020000e+11  7/15/15 16:39     G816     7071096        70201       1.0   
14    2.020000e+11  7/15/15 16:39     G816     7071096        70201       1.0   
15    2.020000e+11  7/15/15 17:11     G816     7000126       210301       1.0   
16    2.020000e+11  7/15/15 17:27      628     7000133       210301       1.0   
17    2.020000e+11  7/15/15 18:06      628     7000300       210101       1.0   
18    2.020000e+11  7/15/15 18:21     G816     7070884        70203       1.0   
19    2.020000e+11  7/15/15 18:21     G816     7071133        70301       1.0   

    price  total  
0     2.0    2.0  
1     2.0    2.0  
2     2.0    2.0  
3     6.0    6.0  
4     6.0    6.0  
5     2.0    2.0  
6     2.0    2.0  
7     3.5    3.5  
8     2.0    2.0  
9     2.0    2.0  
10    2.0    6.0  
11    2.0    2.0  
12    2.0    2.0  
13    5.0    5.0  
14    5.0    5.0  
15    2.0    2.0  
16    3.0    3.0  
17    2.0    2.0  
18    2.0    2.0  
19    2.0    2.0

### g) Sort the data by 'total'. What is the average price of the three most expensive purchases?  <br> (Mark: 2)

In [132]:
df_data.sort_values('total', ascending=False)

invoice_number   invoice_time store_id  product_id  category_id  \
4510    2.020000e+11  8/30/15 20:21      628    21000362       210301   
3784    2.020000e+11  8/26/15 17:12     6666     7000294       210101   
4288    2.020000e+11   8/29/15 8:31     6666     7000294       210101   
4122    2.020000e+11  8/28/15 11:42     6666     7090097        70301   
4876    2.020000e+11   9/1/15 21:20     G816     7000300       210101   
...              ...            ...      ...         ...          ...   
3989    2.020000e+11  8/27/15 15:52     6666     7090455        70101   
3991    2.020000e+11  8/27/15 15:52     6666     7090509        70101   
3059    2.020000e+11  8/21/15 17:59     6666     7000294       210101   
3762    2.020000e+11  8/26/15 15:47     6666     7000294       210101   
3119    2.020000e+11   8/22/15 0:42     6666     7000294       210101   

      num_item  price   total  
4510     116.0    3.0  348.00  
3784      48.0    1.5   72.00  
4288      48.0    1.5   72.00  
4122      20.0    3.0   60.00  
4876      24.0    2.0   48.00  
...        ...    ...     ...  
3989       0.5    1.5    0.75  
3991       0.5    1.5    0.75  
3059       1.0    0.0    0.00  
3762       1.0    0.0    0.00  
3119       1.0    0.0    0.00  

[5382 rows x 8 columns]

In [148]:
df_data.total.nlargest(3).mean()

164.0

### h) First, covert the column invoice time to a datetime column. Print out the new column types to make sure this was done correctly. Next, let's create columns called "DOW" and "Month", which store the corresponding weekday and month of the invoice respectively.  Print the head of the dataframe to make sure this was done correctly. <br> (Mark: 2)

In [134]:
df_data['Time'] = pd.to_datetime(df_data.invoice_time)
df_data.head()

invoice_number   invoice_time store_id  product_id  category_id  num_item  \
0    2.020000e+11  7/15/15 10:45     G816     7000300       210101       1.0   
1    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
2    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
3    2.020000e+11  7/15/15 12:16     G816     7070891        70401       1.0   
4    2.020000e+11  7/15/15 12:16     G816     7090363        70401       1.0   

   price  total                Time  
0    2.0    2.0 2015-07-15 10:45:00  
1    2.0    2.0 2015-07-15 12:16:00  
2    2.0    2.0 2015-07-15 12:16:00  
3    6.0    6.0 2015-07-15 12:16:00  
4    6.0    6.0 2015-07-15 12:16:00

In [135]:
df_data['DOW'] = df_data.Time.dt.day_name()

df_data['Month'] = df_data.Time.dt.month

df_data.head()

invoice_number   invoice_time store_id  product_id  category_id  num_item  \
0    2.020000e+11  7/15/15 10:45     G816     7000300       210101       1.0   
1    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
2    2.020000e+11  7/15/15 12:16     G816     7070884        70203       1.0   
3    2.020000e+11  7/15/15 12:16     G816     7070891        70401       1.0   
4    2.020000e+11  7/15/15 12:16     G816     7090363        70401       1.0   

   price  total                Time        DOW  Month  
0    2.0    2.0 2015-07-15 10:45:00  Wednesday      7  
1    2.0    2.0 2015-07-15 12:16:00  Wednesday      7  
2    2.0    2.0 2015-07-15 12:16:00  Wednesday      7  
3    6.0    6.0 2015-07-15 12:16:00  Wednesday      7  
4    6.0    6.0 2015-07-15 12:16:00  Wednesday      7

### i) How many different product_id are there? <br> (Mark: 1)

In [136]:
df_data.product_id.nunique()

53

### j) Which store is the highest total? <br> (Mark: 1)

In [137]:
df_data.groupby(['store_id'])['total'].sum().nlargest(1)

store_id
G816    8020.0
Name: total, dtype: float64

# Q2

Read the Titanic dataset.

In [138]:
df_titanic = pd.read_csv("Titanic.csv")

df_titanic.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

### a) Change the index of the df_titanic to the column PasengerID. Select the name of passenger 896 using loc. <br> (Mark: 1)

In [139]:
df_titanic.set_index("PassengerId")

Pclass                                          Name     Sex  \
PassengerId                                                                 
892               3                              Kelly, Mr. James    male   
893               3              Wilkes, Mrs. James (Ellen Needs)  female   
894               2                     Myles, Mr. Thomas Francis    male   
895               3                              Wirz, Mr. Albert    male   
896               3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   
...             ...                                           ...     ...   
1305              3                            Spector, Mr. Woolf    male   
1306              1                  Oliva y Ocana, Dona. Fermina  female   
1307              3                  Saether, Mr. Simon Sivertsen    male   
1308              3                           Ware, Mr. Frederick    male   
1309              3                      Peter, Master. Michael J    male   

              Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
PassengerId                                                                   
892          34.5      0      0              330911    7.8292   NaN        Q  
893          47.0      1      0              363272    7.0000   NaN        S  
894          62.0      0      0              240276    9.6875   NaN        Q  
895          27.0      0      0              315154    8.6625   NaN        S  
896          22.0      1      1             3101298   12.2875   NaN        S  
...           ...    ...    ...                 ...       ...   ...      ...  
1305          NaN      0      0           A.5. 3236    8.0500   NaN        S  
1306         39.0      0      0            PC 17758  108.9000  C105        C  
1307         38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
1308          NaN      0      0              359309    8.0500   NaN        S  
1309          NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 10 columns]

### b) How many missing entries are there in the Age column? <br> (Mark: 1)

In [140]:
df_titanic.Age.isnull().sum()

86

### c) Using the fillna() method replace the missing values in the Age column with the mean. <br> (Mark: 1)

In [142]:
df_titanic['Age'].fillna(value=df_titanic['Age'].mean(), inplace=True)


### d) What is the average age of the 5 oldest passengers? <br> (Mark: 2)

In [143]:
df_titanic.sort_values('Age', ascending=False)


PassengerId  Pclass                                               Name  \
96           988       1  Cavendish, Mrs. Tyrell William (Julia Florence...   
81           973       1                                 Straus, Mr. Isidor   
236         1128       1                           Warren, Mr. Frank Manley   
179         1071       1  Compton, Mrs. Alexander Taylor (Mary Eliza Ing...   
305         1197       1  Crosby, Mrs. Edward Gifford (Catherine Elizabe...   
..           ...     ...                                                ...   
250         1142       2                              West, Miss. Barbara J   
307         1199       3                          Aks, Master. Philip Frank   
281         1173       3                     Peacock, Master. Alfred Edward   
201         1093       3            Danbom, Master. Gilbert Sigvard Emanuel   
354         1246       3            Dean, Miss. Elizabeth Gladys Millvina""   

        Sex    Age  SibSp  Parch              Ticket      Fare    Cabin  \
96   female  76.00      1      0               19877   78.8500      C46   
81     male  67.00      1      0            PC 17483  221.7792  C55 C57   
236    male  64.00      1      0              110813   75.2500      D37   
179  female  64.00      0      2            PC 17756   83.1583      E45   
305  female  64.00      1      1              112901   26.5500      B26   
..      ...    ...    ...    ...                 ...       ...      ...   
250  female   0.92      1      2          C.A. 34651   27.7500      NaN   
307    male   0.83      0      1              392091    9.3500      NaN   
281    male   0.75      1      1  SOTON/O.Q. 3101315   13.7750      NaN   
201    male   0.33      0      2              347080   14.4000      NaN   
354  female   0.17      1      2           C.A. 2315   20.5750      NaN   

    Embarked  
96         S  
81         S  
236        C  
179        C  
305        S  
..       ...  
250        S  
307        S  
281        S  
201        S  
354        S  

[418 rows x 11 columns]

In [150]:
df_titanic.Age.nlargest(5).mean()

67.0